sda/cmd/api/api.go

```diff
@@ -344,6 +354,205 @@ func deleteFile(c *gin.Context) {
+	fileID = strings.TrimPrefix(fileID, "/")
+
+	if fileID == "" {
+		c.AbortWithStatusJSON(http.StatusBadRequest, "file ID is required")
+
+		return
+	}
+
+	// Retrieve the actual file path for the user's file.
+	filePath, err := Conf.API.DB.GetInboxFilePathFromID(submissionUser, fileID, false)
+	if err != nil {
+		log.Errorf("getting file path from fileID (%s) failed, reason: (%v)", fileID, err)
+		c.AbortWithStatusJSON(http.StatusNotFound, "File could not be found in the inbox")
+
+		return
+	}
+
+	// Get inbox file handle
+	file, err := inbox.NewFileReader(filePath)
+	if err != nil {
+		log.Errorf("could not find inbox file %s, %s", filePath, err)
+		c.String(http.StatusInternalServerError, "inbox error")
+
+		return
+	}
+
+	// get the header of the crypt4gh file
+	headerConsumed := false
+	header, err := Conf.API.DB.GetHeader(fileID)
+	if err != nil {
+		log.Debugf("failed to get header for fileID %s from the database, reason: %v", fileID, err)
+		log.Debugln("trying to read header from the file")
+
+		header, err = headers.ReadHeader(file)
+		headerConsumed = true
+		if err != nil {
+			log.Debugf("failed to read header for fileID %s, reason: %v", fileID, err)
+			c.AbortWithStatusJSON(http.StatusInternalServerError, err.Error())
+
+			return
+		}
+	}
+
+	decryptedChecksum, err := Conf.API.DB.GetDecryptedChecksum(fileID)
+	if err != nil {
+		log.Debugf("Warning: failed to get decrypted checksum for fileID %s, reason: %v", fileID, err)
+	}
+
+	// Get the public key from the request header.
+	reEncKey := c.GetHeader("Client-Public-Key")
+	if reEncKey == "" {
+		c.String(http.StatusBadRequest, "client public key is not provided in the header")
+
+		return
+	}
+
+	// Set the headers for the response.
+	c.Header("Content-Type", "application/octet-stream")
+	c.Header("Content-Disposition", fmt.Sprintf("filename: %v", fileID))
+	c.Header("ETag", fmt.Sprintf("decryptedChecksum: %s", decryptedChecksum))
+
+	newHeader, err := reencryptHeaderFunc(header, reEncKey)
+	if err != nil {
+		log.Errorf("failed to reencrypt header, reason: %v", err)
+		c.String(http.StatusInternalServerError, "failed to reencrypt header")
+
+		return
+	}
+
+	// Skip the original header in the file if it hasn't been consumed yet.
+	if !headerConsumed {
+		headerLength := int64(len(header))
+		_, err = io.CopyN(io.Discard, file, headerLength)
+		if err != nil {
+			log.Errorf("failed to skip original header, reason: %v", err)
+			c.String(http.StatusInternalServerError, "failed to skip original header")
+
+			return
+		}
+	}
+
+	fileStream := io.MultiReader(bytes.NewReader(newHeader), file)
+
+	err = sendStream(c.Writer, fileStream)
+	if err != nil {
+		log.Errorf("error occurred while sending stream: %v", err)
+		c.String(http.StatusInternalServerError, "an error occurred")
+
+		return
+	}
+
+	c.Status(http.StatusOK)
+}
+
+// sendStream streams file contents from a reader
+var sendStream = func(w http.ResponseWriter, file io.Reader) error {
```

Comment:
This function should have  a separate test.

Reply:
fixed in 1fcc13f3